packages/askar/src/AskarModule.ts

```diff
@@ -16,52 +16,68 @@ export class AskarModule implements Module {
-      }
+    if (this.config.enableKms) {
+      const kmsConfig = dependencyManager.resolve(Kms.KeyManagementModuleConfig)
+      kmsConfig.registerBackend(new AksarKeyManagementService())
     }
 
-    if (dependencyManager.isRegistered(InjectionSymbols.StorageService)) {
-      throw new CredoError('There is an instance of StorageService already registered')
-    } else {
+    if (this.config.enableStorage) {
+      if (dependencyManager.isRegistered(InjectionSymbols.StorageService)) {
+        throw new CredoError(
+          'Unable to register AskatStoreService. There is an instance of StorageService already registered'
+        )
+      }
       dependencyManager.registerSingleton(InjectionSymbols.StorageService, AskarStorageService)
     }
+
+    dependencyManager.registerSingleton(AskarStoreManager)
   }
 
-  public async initialize(agentContext: AgentContext): Promise<void> {
-    // We MUST use an askar wallet here
-    assertAskarWallet(agentContext.wallet)
-
-    const wallet = agentContext.wallet
-
-    // Register the Askar store instance on the dependency manager
-    // This allows it to be re-used for tenants
-    agentContext.dependencyManager.registerInstance(Store, agentContext.wallet.store)
-
-    // If the multiWalletDatabaseScheme is set to ProfilePerWallet, we want to register the AskarProfileWallet
-    // and return that as the wallet for all tenants, but not for the main agent, that should use the AskarWallet
-    if (this.config.multiWalletDatabaseScheme === AskarMultiWalletDatabaseScheme.ProfilePerWallet) {
-      agentContext.dependencyManager.container.register(InjectionSymbols.Wallet, {
-        useFactory: (container) => {
-          // If the container is the same as the root dependency manager container
-          // it means we are in the main agent, and we should use the root wallet
-          if (container === agentContext.dependencyManager.container) {
-            return wallet
-          }
-
-          // Otherwise we want to return the AskarProfileWallet
-          return container.resolve(AskarProfileWallet)
-        },
-      })
+  public async onInitializeContext(agentContext: AgentContext, metadata: Record<string, unknown>) {
+    // TODO: I think we should also register the store here
+    if (agentContext.contextCorrelationId === 'default') {
+      const storeManager = agentContext.dependencyManager.resolve(AskarStoreManager)
+      await storeManager.getInitializedStoreWithProfile(agentContext)
+      return
+    } else if (this.config.multiWalletDatabaseScheme === AskarMultiWalletDatabaseScheme.DatabasePerWallet) {
+      agentContext.dependencyManager.registerInstance('AskarContextMetadata', metadata)
+      const storeManager = agentContext.dependencyManager.resolve(AskarStoreManager)
+      await storeManager.getInitializedStoreWithProfile(agentContext)
     }
   }
+
+  public async onProvisionContext(agentContext: AgentContext) {
```

Comment:
I added new module methods:
- onProvisionContext
- onCloseContext
- onDeleteContext
- onInitializeContext

this will allow for lifecycle handling of specific contexts

Reply:
This is really great!